File: simpleast-core/src/main/java/com/agarron/simpleast_core/utils/ASTUtils.java
package com.agarron.simpleast_core.utils;

import com.agarron.simpleast_core.node.Node;

import java.util.Collection;
import java.util.List;

public class ASTUtils {
    public static void traversePostOrder(final Collection<? extends Node> ast, final NodeProcessor nodeProcessor) {
        for (final Node node : ast) {
            traversePostOrderSubtree(node, nodeProcessor);
        }
    }

    private static void traversePostOrderSubtree(final Node node, final NodeProcessor nodeProcessor) {
        if (node.hasChildren()) {
            final List<Node> children = node.getChildren();
            for (final Node child : children) {
                traversePostOrderSubtree(child, nodeProcessor);
            }
        }

        nodeProcessor.processNode(node);
    }
}

File: simpleast-core/src/main/java/com/agarron/simpleast_core/ASTUtils.kt
package com.agarron.simpleast_core

import com.agarron.simpleast_core.Node
import com.agarron.simpleast_core.Parent

class ASTUtils {
    companion object {
        @JvmStatic
        fun processTree(ast: Collection<Node>, nodeProcessor: (Node) -> Unit) {
            ast.forEach {
                processSubtree(it, nodeProcessor)
            }
        }

        private fun processSubtree(node: Node, nodeProcessor: (Node) -> Unit) {
            if (node is Parent) {
                node.getChildren().forEach { processSubtree(it, nodeProcessor) }
            }

            nodeProcessor.invoke(node)
        }
    }
}
